File: app/dashboard/drone-dashboard/src/constants/px4FlightModes.js
/**
 * PX4 Flight Mode Constants - Official Standards Implementation
 *
 * Based on official PX4-Autopilot source code v1.15+
 * References:
 * - PX4-Autopilot/src/modules/commander/px4_custom_mode.h (Official definitions)
 * - MAVLink common.xml protocol specification
 * - PX4 Developer Guide flight mode documentation
 * - Tested against real PX4 HEARTBEAT messages
 *
 * PX4 stores custom_mode as a little-endian union:
 * - bits 0..15: reserved
 * - bits 16..23: main_mode
 * - bits 24..31: sub_mode
 *
 * Therefore an AUTO mission heartbeat is 0x04040000, not 0x00040004.
 */

/**
 * Official PX4 Main Mode Definitions (bits 16..23 of custom_mode)
 * Source: PX4-Autopilot/src/modules/commander/px4_custom_mode.h
 */
export const PX4_MAIN_MODES = {
  MANUAL: 1,
  ALTCTL: 2,
  POSCTL: 3,
  AUTO: 4,
  ACRO: 5,
  OFFBOARD: 6,
  STABILIZED: 7,
  RATTITUDE: 8
};

/**
 * Official PX4 Auto Sub-Mode Definitions (bits 24..31 when main_mode = AUTO)
 * Source: PX4-Autopilot/src/modules/commander/px4_custom_mode.h
 */
export const PX4_AUTO_SUB_MODES = {
  AUTO_READY: 1,
  AUTO_TAKEOFF: 2,
  AUTO_LOITER: 3,
  AUTO_MISSION: 4,
  AUTO_RTL: 5,
  AUTO_LAND: 6,
  AUTO_RESERVED_DO_NOT_USE: 7,
  AUTO_FOLLOW_TARGET: 8,
  AUTO_PRECLAND: 9
};

const UINT32_MAX = 0xFFFFFFFF;

const normalizeUint32 = (value) => {
  if (typeof value === 'string' && value.trim() === '') {
    return null;
  }

  const numericValue = typeof value === 'string' ? Number(value) : value;
  if (!Number.isInteger(numericValue) || numericValue < 0 || numericValue > UINT32_MAX) {
    return null;
  }

  return numericValue >>> 0;
};

/**
 * Encode the two PX4 mode bytes into MAVLink HEARTBEAT.custom_mode.
 *
 * @param {number} mainMode - PX4_CUSTOM_MAIN_MODE value
 * @param {number} [subMode=0] - PX4_CUSTOM_SUB_MODE value
 * @returns {number} Unsigned 32-bit PX4 custom_mode
 */
export const encodePx4CustomMode = (mainMode, subMode = 0) => {
  if (
    !Number.isInteger(mainMode) || mainMode < 0 || mainMode > 0xFF ||
    !Number.isInteger(subMode) || subMode < 0 || subMode > 0xFF
  ) {
    throw new RangeError('PX4 main mode and sub-mode must be unsigned bytes');
  }

  return (((subMode & 0xFF) << 24) | ((mainMode & 0xFF) << 16)) >>> 0;
};

/**
 * Decode MAVLink HEARTBEAT.custom_mode using PX4's px4_custom_mode union.
 *
 * @param {number|string} customMode - Unsigned PX4 custom_mode value
 * @returns {{customMode: number, reserved: number, mainMode: number, subMode: number}|null}
 */
export const decodePx4CustomMode = (customMode) => {
  const mode = normalizeUint32(customMode);
  if (mode === null) {
    return null;
  }

  return {
    customMode: mode,
    reserved: mode & 0xFFFF,
    mainMode: (mode >>> 16) & 0xFF,
    subMode: (mode >>> 24) & 0xFF
  };
};

const AUTO_SUB_MODE_NAMES = {
  [PX4_AUTO_SUB_MODES.AUTO_READY]: 'Ready',
  [PX4_AUTO_SUB_MODES.AUTO_TAKEOFF]: 'Takeoff',
  [PX4_AUTO_SUB_MODES.AUTO_LOITER]: 'Hold',
  [PX4_AUTO_SUB_MODES.AUTO_MISSION]: 'Mission',
  [PX4_AUTO_SUB_MODES.AUTO_RTL]: 'Return',
  [PX4_AUTO_SUB_MODES.AUTO_LAND]: 'Land',
  [PX4_AUTO_SUB_MODES.AUTO_RESERVED_DO_NOT_USE]: 'Auto Reserved',
  [PX4_AUTO_SUB_MODES.AUTO_FOLLOW_TARGET]: 'Follow Target',
  [PX4_AUTO_SUB_MODES.AUTO_PRECLAND]: 'Precision Land'
};

/**
 * Official PX4 Flight Mode Mappings.
 * Keys are generated from PX4's byte layout to keep the table and decoder aligned.
 */
export const PX4_FLIGHT_MODES = {
  // System States
  0: 'Initializing',

  // Manual Control Modes
  [encodePx4CustomMode(PX4_MAIN_MODES.MANUAL)]: 'Manual',
  [encodePx4CustomMode(PX4_MAIN_MODES.ALTCTL)]: 'Altitude',
  [encodePx4CustomMode(PX4_MAIN_MODES.POSCTL)]: 'Position',
  [encodePx4CustomMode(PX4_MAIN_MODES.ACRO)]: 'Acro',
  [encodePx4CustomMode(PX4_MAIN_MODES.OFFBOARD)]: 'Offboard',
  [encodePx4CustomMode(PX4_MAIN_MODES.STABILIZED)]: 'Stabilized',
  [encodePx4CustomMode(PX4_MAIN_MODES.RATTITUDE)]: 'Rattitude',

  // Auto Modes with Sub-modes
  [encodePx4CustomMode(PX4_MAIN_MODES.AUTO)]: 'Auto',
  ...Object.fromEntries(
    Object.entries(AUTO_SUB_MODE_NAMES).map(([subMode, name]) => [
      encodePx4CustomMode(PX4_MAIN_MODES.AUTO, Number(subMode)),
      name
    ])
  )
};

/**
 * Official MAVLink System Status (MAV_STATE) enumeration
 * Source: MAVLink common.xml specification
 */
export const MAV_STATE = {
  0: 'Uninit',
  1: 'Boot',
  2: 'Calibrating',
  3: 'Standby',
  4: 'Active',
  5: 'Critical',
  6: 'Emergency',
  7: 'Poweroff',
  8: 'Flight Termination'
};

/**
 * Standards-compliant flight mode detection with PX4 decoding
 * @param {number} customMode - PX4 custom_mode from HEARTBEAT message
 * @returns {string} Human-readable flight mode name
 */
export const getFlightModeName = (customMode) => {
  if (customMode === null || customMode === undefined) {
    return 'No Data';
  }

  const decoded = decodePx4CustomMode(customMode);
  if (!decoded) {
    return 'Invalid Mode';
  }

  const { customMode: mode, mainMode, subMode } = decoded;

  // Direct lookup for known modes
  if (Object.prototype.hasOwnProperty.call(PX4_FLIGHT_MODES, mode)) {
    return PX4_FLIGHT_MODES[mode];
  }

  // Handle main modes with unknown sub-modes
  switch (mainMode) {
    case PX4_MAIN_MODES.MANUAL:
      return 'Manual';
    case PX4_MAIN_MODES.ALTCTL:
      return 'Altitude';
    case PX4_MAIN_MODES.POSCTL:
      return 'Position';
    case PX4_MAIN_MODES.AUTO:
      return AUTO_SUB_MODE_NAMES[subMode] || (subMode === 0 ? 'Auto' : `Auto.${subMode}`);
    case PX4_MAIN_MODES.ACRO:
      return 'Acro';
    case PX4_MAIN_MODES.OFFBOARD:
      return 'Offboard';
    case PX4_MAIN_MODES.STABILIZED:
      return 'Stabilized';
    case PX4_MAIN_MODES.RATTITUDE:
      return 'Rattitude';

    default:
      return `Unknown (${mode})`;
  }
};

/**
 * Get system status name from MAV_STATE value
 * @param {number} systemStatus - MAV_STATE from HEARTBEAT message
 * @returns {string} Human-readable system status
 */
export const getSystemStatusName = (systemStatus) => {
  if (systemStatus === null || systemStatus === undefined) {
    return 'No Data';
  }

  const status = typeof systemStatus === 'string' ? parseInt(systemStatus, 10) : systemStatus;

  if (isNaN(status)) {
    return 'Invalid Status';
  }

  return MAV_STATE[status] || `Unknown (${status})`;
};

/**
 * Check if flight mode allows manual pilot control
 * @param {number} customMode - PX4 custom_mode value
 * @returns {boolean} True if manual control available
 */
export const isSafeFlightMode = (customMode) => {
  const decoded = decodePx4CustomMode(customMode);
  if (!decoded) return false;

  // Modes with direct pilot control are considered "safe" by this UI helper.
  return [
    PX4_MAIN_MODES.MANUAL,
    PX4_MAIN_MODES.ALTCTL,
    PX4_MAIN_MODES.POSCTL,
    PX4_MAIN_MODES.ACRO,
    PX4_MAIN_MODES.STABILIZED,
    PX4_MAIN_MODES.RATTITUDE
  ].includes(decoded.mainMode);
};

/**
 * Check if system is ready for operations
 * @param {number} systemStatus - MAV_STATE value
 * @returns {boolean} True if system ready
 */
export const isSystemReady = (systemStatus) => {
  const status = typeof systemStatus === 'string' ? parseInt(systemStatus, 10) : systemStatus;

  if (isNaN(status)) return false;

  // Standby (3) and Active (4) are operational states
  return status === 3 || status === 4;
};

/**
 * Get flight mode category for UI styling
 * @param {number} customMode - PX4 custom_mode value
 * @returns {string} Category: 'manual', 'auto', 'offboard', or 'unknown'
 */
export const getFlightModeCategory = (customMode) => {
  const decoded = decodePx4CustomMode(customMode);
  if (!decoded) return 'unknown';

  switch (decoded.mainMode) {
    case PX4_MAIN_MODES.MANUAL:
    case PX4_MAIN_MODES.ALTCTL:
    case PX4_MAIN_MODES.POSCTL:
    case PX4_MAIN_MODES.ACRO:
    case PX4_MAIN_MODES.STABILIZED:
    case PX4_MAIN_MODES.RATTITUDE:
      return 'manual';

    case PX4_MAIN_MODES.AUTO:
      return 'auto';

    case PX4_MAIN_MODES.OFFBOARD:
      return 'offboard';

    default:
      return 'unknown';
  }
};

File: src/px4_flight_modes.py
"""PX4 ``HEARTBEAT.custom_mode`` decoding shared by node runtime code.

PX4 stores a reserved uint16 followed by one byte each for ``main_mode`` and
``sub_mode``.  Keeping this protocol knowledge here prevents mission scripts
and telemetry processing from growing independent numeric lookup tables.
"""

PX4_CUSTOM_MAIN_MODE_SHIFT = 16
PX4_CUSTOM_SUB_MODE_SHIFT = 24
PX4_CUSTOM_MODE_BYTE_MASK = 0xFF

PX4_MAIN_MODE_MANUAL = 1
PX4_MAIN_MODE_ALTCTL = 2
PX4_MAIN_MODE_POSCTL = 3
PX4_MAIN_MODE_AUTO = 4
PX4_MAIN_MODE_ACRO = 5
PX4_MAIN_MODE_OFFBOARD = 6
PX4_MAIN_MODE_STABILIZED = 7
PX4_MAIN_MODE_RATTITUDE_LEGACY = 8
PX4_MAIN_MODE_SIMPLE = 9
PX4_MAIN_MODE_TERMINATION = 10
PX4_MAIN_MODE_ALTITUDE_CRUISE = 11

PX4_MAIN_MODE_NAMES = {
    PX4_MAIN_MODE_MANUAL: "Manual",
    PX4_MAIN_MODE_ALTCTL: "Altitude",
    PX4_MAIN_MODE_ACRO: "Acro",
    PX4_MAIN_MODE_OFFBOARD: "Offboard",
    PX4_MAIN_MODE_STABILIZED: "Stabilized",
    PX4_MAIN_MODE_RATTITUDE_LEGACY: "Rattitude",
    PX4_MAIN_MODE_SIMPLE: "Simple",
    PX4_MAIN_MODE_TERMINATION: "Termination",
    PX4_MAIN_MODE_ALTITUDE_CRUISE: "Altitude Cruise",
}

PX4_POSCTL_SUB_MODE_NAMES = {
    0: "Position",
    1: "Orbit",
    2: "Position Slow",
}

PX4_AUTO_SUB_MODE_NAMES = {
    0: "Auto",
    1: "Ready",
    2: "Takeoff",
    3: "Hold",
    4: "Mission",
    5: "Return",
    6: "Land",
    7: "Auto Reserved",
    8: "Follow Target",
    9: "Precision Land",
    10: "VTOL Takeoff",
    11: "External 1",
    12: "External 2",
    13: "External 3",
    14: "External 4",
    15: "External 5",
    16: "External 6",
    17: "External 7",
    18: "External 8",
    19: "Guided Course",
    20: "Descend",
}


def decode_px4_custom_mode(custom_mode: int | None) -> tuple[int, int]:
    """Return the PX4 ``(main_mode, sub_mode)`` bytes from custom_mode."""
    mode = int(custom_mode or 0)
    main_mode = (mode >> PX4_CUSTOM_MAIN_MODE_SHIFT) & PX4_CUSTOM_MODE_BYTE_MASK
    sub_mode = (mode >> PX4_CUSTOM_SUB_MODE_SHIFT) & PX4_CUSTOM_MODE_BYTE_MASK
    return main_mode, sub_mode


def describe_px4_custom_mode(custom_mode: int | None) -> str:
    """Return a stable operator-facing name for a raw PX4 custom_mode."""
    mode = int(custom_mode or 0)
    if mode == 0:
        return "Unknown/Uninit"

    main_mode, sub_mode = decode_px4_custom_mode(mode)
    if main_mode == PX4_MAIN_MODE_POSCTL:
        return PX4_POSCTL_SUB_MODE_NAMES.get(sub_mode, f"Position({sub_mode})")
    if main_mode == PX4_MAIN_MODE_AUTO:
        return PX4_AUTO_SUB_MODE_NAMES.get(sub_mode, f"Auto({sub_mode})")
    if main_mode in PX4_MAIN_MODE_NAMES:
        return PX4_MAIN_MODE_NAMES[main_mode]
    return f"Unknown({mode})"
